futhark: init at 0.6.2

infinisil opened the pull request:
###### Motivation for this change
Predicate to be able to demonstrate how well Nix can work as Futhark's package manager for https://www.reddit.com/r/ProgrammingLanguages/comments/90fnop/the_future_futhark_package_manager/

Edit: It's live :) https://github.com/infinisil/futharknixdemo

Tested with a simple compilation and run. It needed some fancy gcc `NIX_` stuff to work, @cleverca22 helped a lot with that on IRC :).

It's a Haskell package not present in hackage. I used cabal2nix to generate a file with the dependencies and made the updating process automated via a small script.

Ping @peti @domenkozar for Haskell

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

infinisil committed:
futhark: init at 0.6.2

Infinisil commented on pkgs/development/compilers/futhark/package.nix:
```diff
@@ -0,0 +1,49 @@
+mkDerivation {
+  pname = "futhark";
+  version = "0.6.2";
+  src = /nix/store/5n2kpq21b3hm0yfr7wsinh5v24y81s98-source;
```
While a bit ugly, this is always overridden by the `default.nix`.

mpickering replied:
Why not take the `src` as an argument? This is too ugly.

Infinisil replied:
This is autogenerated by cabal2nix, I can't just change what it generates. I also don't want to have to manually edit the file every time an update comes out. It would be possible to use `cabal2nix`'s git fetching. But that has the problem that it doesn't use the github tarballs, it fetches all submodules, and futhark has a super big submodule that takes forever to fetch. This may get better with https://github.com/NixOS/cabal2nix/issues/260. I don't know of a cleaner solution than this.

mpickering commented on pkgs/development/compilers/futhark/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  path = stdenv.lib.makeBinPath [ gcc ];
+
+  wrapped = symlinkJoin {
```
Why do you use `symlinkJoin` with one one thing in `paths`?

Infinisil replied:
Makes it easy to patch a single file, since everything gets symlinked and I can just replace a symlink. Idris also uses this in its wrapper.

Alternative would be to do `ln -s ${package}/* $out; rm $out/bin; mkdir $out/bin; ln -s ${package}/bin/* $out/bin; rm $out/bin/futhark-c; rm $out/bin/futhark-opencl` and then the makeWrappers. I can do that if you prefer.

mpickering commented on pkgs/development/compilers/futhark/update:
```diff
@@ -0,0 +1,15 @@
+	exit 1
+fi
+
+echo Updated package.nix file successfully
```
I feel like if you have to use this script then it should be packaged properly and automatically run on whatever `src` you choose rather than just living in the directory.

Infinisil commented:
@mpickering I removed the update script and replaced it with a short manual description of how to update in the file, is this alright?

mpickering commented:
Where is the description?

mpickering commented:
OK, I think this is fine. It seems annoying that there isn't one invocation of `cabal2nix` which will generate this file but I can't get it to work.

For example:

```
cabal2nix https://github.com/diku-dk/futhark --hpack --revision v0.6.2
```

Infinisil commented:
Yeah currently there isn't, as previously mentioned, there is an issue for it: https://github.com/NixOS/cabal2nix/issues/260

Infinisil commented:
@peti @basvandijk Is it okay to add a haskell package like this? It's not in hackage

flokli commented:
Was able to compile some examples with it. Unfortunately couldn't execute them, as I couldn't get OpenCL to work on my hardware (yet).

`opencl-info` complains
```
ERROR: clGetPlatformIDs(-1001)
```

Thinkpad T460p with i7-6820HQ, currently not using the nvidia card `NVIDIA Corporation GM108M [GeForce 940MX]` at all. According to [this](https://en.wikichip.org/wiki/intel/core_i7/i7-6820hq), the Intel Graphics should support OpenCL on it's own, so not really sure what do do here to get it working `¯\_(ツ)_/¯`

basvandijk committed:
futhark: introduce the top-level futhark attribute

The fact that futhark is a Haskell package is an implementation detail. To
install it users should just have to specify `futhark` instead of
`haskellPackages.futhark`.

Additionally futhark is overridden with `haskell.lib.justStaticExecutables` to
reduce closure size.

basvandijk reviewed (changes requested):
@Infinisil cool compiler!

The fact that futhark is a Haskell package is an implementation detail. To install it users should just have to specify `futhark` instead of `haskellPackages.futhark`.

Note that we also do this for most other Haskell executables. Like those other Haskell executables we should also override futhark with `haskell.lib.justStaticExecutables` to
reduce closure size.

I made these changes in: https://github.com/LumiGuide/nixpkgs/commit/5ba57d552090eeb029ff2bfd6ddd707542420902. You may want to merge that into your branch or I can also do it if you like.

Offtopic: note that in https://github.com/NixOS/nixpkgs/pull/45406 I'm disabling library profiling for `justStaticExecutables` meaning that once that is merged `futhark` will build much quicker.

Infinisil commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -510,4 +510,20 @@ self: super: builtins.intersectAttrs super {
+        wrapProgram $out/bin/futhark-c \
+          --prefix PATH : "${path}"
+
+        wrapProgram $out/bin/futhark-opencl \
```
@basvandijk These wrapper don't seem to be applied like this, building either `pkgs.futhark` or `pkgs.haskellPackages.futhark` results in raw ELF binaries

basvandijk replied:
Strange, I did test if it created the correct wrappers but I can confirm it now results in ELF binaries. I'll look into it.

basvandijk replied:
This should now be fixed by https://github.com/NixOS/nixpkgs/pull/43865/commits/2cdeee2367ccc478b25a6e1db52d099e3d258896 which I pushed to this PR.

@peti OK with this solution?

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=7dd6fb99-7622-4fa3-988a-29888c558977)

The following builds were skipped because they don't evaluate on aarch64-linux: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=fcea2f45-e5ad-4889-bd06-3874e5c6b1fb)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2/bin/futhark-pyopencl
shrinking /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2/lib/ghc-8.4.3/futhark-0.6.2/HSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt.o
wrong ELF type
shrinking /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2/lib/ghc-8.4.3/x86_64-linux-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.so
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2/lib  /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2/bin
patching script interpreter paths in /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2
checking for references to /build in /nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2...
wrong ELF type
/nix/store/fnrkf3pc4gf6bqmgy6n7xmar0fi8pcmm-futhark-0.6.2
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=65ab38e4-3479-46fa-a509-27a7a9b4727b)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib/ghc-8.4.3/x86_64-osx-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.dylib: fixing dylib
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib/ghc-8.4.3/x86_64-osx-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.dylib: fixing dylib
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib/ghc-8.4.3/x86_64-osx-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.dylib: fixing dylib
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib/ghc-8.4.3/x86_64-osx-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.dylib: fixing dylib
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib/ghc-8.4.3/x86_64-osx-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.dylib: fixing dylib
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib/ghc-8.4.3/x86_64-osx-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.dylib: fixing dylib
strip is /nix/store/vvbmqsh5ikrxkiwj71dwfbzcwf4bfav8-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/lib  /nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2/bin
patching script interpreter paths in /nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2
/nix/store/vda5nfqdclskhxcqqrxmh8gysgj28rhp-futhark-0.6.2
```
</p></details>

basvandijk committed:
haskell: have a dedicated file for non Hackage packages

We had a few "overrides" in configuration-common.nix that were really
extensions. They introduced packages that weren't in hackage-packages.nix.

The advantage of having a dedicated file for these packages is that we can still
place Nix-specific overrides to these packages in configuration-nix.nix. We
weren't able do this before because configuration-nix.nix extended only the
packages from hackage-packages.nix.

basvandijk commented:
@GrahamcOfBorg build futhark

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=ca4a4485-aadd-4f21-83f4-16152ac7a789)

The following builds were skipped because they don't evaluate on x86_64-darwin: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=1b30f61b-1d4b-4528-be24-b3adf2962e91)

The following builds were skipped because they don't evaluate on aarch64-linux: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=4cc520b0-fcb5-4322-8636-f313a082a6f6)

The following builds were skipped because they don't evaluate on x86_64-darwin: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=95ec4e4f-c9f1-4888-8e82-a8ee2f2ff696)

The following builds were skipped because they don't evaluate on aarch64-linux: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

basvandijk committed:
haskellPackages.multi-ghc-travis: removed in favour of haskell-ci

multi-ghc-travis has been renamed to haskell-ci which is now also on Hackage.

basvandijk committed:
haskell: add GitHub issues to non-hackage-packages.nix

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=450444a8-7a3a-4a58-9be4-ef25c89f0328)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin/futhark-dataset
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin/.futhark-c-wrapped
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin/futhark-py
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin/futhark-doc
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/lib  /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin
patching script interpreter paths in /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2
checking for references to /build in /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2...
wrong ELF type
/nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2
```
</p></details>

GrahamcOfBorg commented:
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=bde68239-0efc-42c2-a5b0-072cadbc5ff5)

The following builds were skipped because they don't evaluate on x86_64-darwin: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=d0fbfccd-2b77-43df-919f-e834c6640676)

The following builds were skipped because they don't evaluate on aarch64-linux: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=1dcaaaf3-8449-43ef-bc27-d528af8828a6)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin/.futhark-opencl-wrapped
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/lib/ghc-8.4.3/futhark-0.6.2/HSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt.o
wrong ELF type
shrinking /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/lib/ghc-8.4.3/x86_64-linux-ghc-8.4.3/libHSfuthark-0.6.2-FbXK1P8KWFK3P10QxD4Fxt-ghc8.4.3.so
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/lib  /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2/bin
patching script interpreter paths in /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2
checking for references to /build in /nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2...
wrong ELF type
/nix/store/rp6s3l8f5nhd96cndw1yix6hvdqcxyyw-futhark-0.6.2
```
</p></details>

GrahamcOfBorg commented:
Timed out, unknown build status on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=2143e686-9ddf-4062-b304-74a3e014db4d)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
[178 of 191] Compiling Language.Futhark.TypeChecker.Types ( src/Language/Futhark/TypeChecker/Types.hs, dist/build/Language/Futhark/TypeChecker/Types.p_o )
[179 of 191] Compiling Language.Futhark.TypeChecker.Terms ( src/Language/Futhark/TypeChecker/Terms.hs, dist/build/Language/Futhark/TypeChecker/Terms.p_o )
[180 of 191] Compiling Language.Futhark.TypeChecker ( src/Language/Futhark/TypeChecker.hs, dist/build/Language/Futhark/TypeChecker.p_o )
[181 of 191] Compiling Futhark.Compiler.Program ( src/Futhark/Compiler/Program.hs, dist/build/Futhark/Compiler/Program.p_o )
[182 of 191] Compiling Futhark.Compiler.TH ( src/Futhark/Compiler/TH.hs, dist/build/Futhark/Compiler/TH.p_o )
[183 of 191] Compiling Language.Futhark.Futlib.Prelude ( src/Language/Futhark/Futlib/Prelude.hs, dist/build/Language/Futhark/Futlib/Prelude.p_o )
[184 of 191] Compiling Futhark.Internalise.Monomorphise ( src/Futhark/Internalise/Monomorphise.hs, dist/build/Futhark/Internalise/Monomorphise.p_o )
[185 of 191] Compiling Futhark.Internalise ( src/Futhark/Internalise.hs, dist/build/Futhark/Internalise.p_o )
building of '/nix/store/yll2lslq7xg78w57wqcd5vspqygf3r3s-futhark-0.6.2.drv' timed out after 3600 seconds
error: build of '/nix/store/yll2lslq7xg78w57wqcd5vspqygf3r3s-futhark-0.6.2.drv' failed
```
</p></details>

basvandijk commented:
@GrahamcOfBorg build haskell-ci
since I also changed that in this PR.

@Infinisil I'll let you do the honours of merging this in if you think it's OK now.

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=737bca72-17a5-4878-a1ec-5372178b8b51)

The following builds were skipped because they don't evaluate on aarch64-linux: haskell-ci

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

Infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -8423,7 +8425,7 @@ with pkgs;
   msitools = callPackage ../development/tools/misc/msitools { };
 
-  multi-ghc-travis = haskell.lib.justStaticExecutables haskellPackages.multi-ghc-travis;
+  haskell-ci = haskell.lib.justStaticExecutables haskellPackages.haskell-ci;
```
@basvandijk Why this change?

basvandijk replied:
While moving the non-hackage packages to a dedicated file I discovered that `multi-ghc-travis` has been renamed to `haskell-ci` and that `haskell-ci` is now on Hackage. This is explained in: https://github.com/NixOS/nixpkgs/commit/6106664ffab6dd5f6f2c1191969545d6b4f88ee1

basvandijk replied:
Mmm maybe an alias might be in order here...

GrahamcOfBorg commented:
<!--REQUEST_ID=4a8407d3-7d0a-4e71-83b0-5a8a09f27f03-->
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=cf5ebe96-a2a1-4129-9d76-feece6a2271d)

The following builds were skipped because they don't evaluate on x86_64-darwin: futhark

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

Infinisil commented:
Rebased to fix conflicts, tested the binary again (still works) and added a `multi-ghc-travis` with an explanatory error message.

GrahamcOfBorg commented:
<!--REQUEST_ID=4a8407d3-7d0a-4e71-83b0-5a8a09f27f03-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=e00508d2-fe2d-4752-abe8-72999981e36b)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2/bin/futhark-dataset
shrinking /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2/bin/.futhark-c-wrapped
shrinking /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2/bin/futhark-py
shrinking /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2/bin/futhark-doc
strip is /nix/store/dxf1m7dhc4qb655bdljc1fsd74v1nag3-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2/lib  /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2/bin
patching script interpreter paths in /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2
checking for references to /build in /nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2...
wrong ELF type
/nix/store/hfisnq5n4kwaajdksm8jlacwkij5riz9-futhark-0.6.2
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=4a8407d3-7d0a-4e71-83b0-5a8a09f27f03-->
Timed out, unknown build status on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43865&attempt_id=4b9dba07-8c64-4f50-be0d-872527c1eeb1)

Attempted: futhark

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/k4qv9vv8lnvyny08diixxrhqzfl9711f-tls-1.4.1.drv': 17 dependencies couldn't be built
cannot build derivation '/nix/store/5i16y4qpsvxw3z86hgvr4342kr47amdv-conduit-extra-1.3.0.drv': 16 dependencies couldn't be built
cannot build derivation '/nix/store/37ak50dqw2vf2h0jmvkh1zcb3s6rs0qz-connection-0.2.8.drv': 11 dependencies couldn't be built
cannot build derivation '/nix/store/kbcbj0abxbjwx836497pwi74bzc0zgp9-yaml-0.8.32.drv': 15 dependencies couldn't be built
cannot build derivation '/nix/store/r9dgp46g10yrgc29xb8rn3x3g6ksa5vl-http-client-tls-0.3.5.3.drv': 13 dependencies couldn't be built
cannot build derivation '/nix/store/has4npvwn5343b4idn4w4m537ip3ny88-xml-conduit-1.8.0.1.drv': 13 dependencies couldn't be built
cannot build derivation '/nix/store/kmp3mnp5lgqvzgv740gc2n2b2v6096f4-hpack-0.28.2.drv': 19 dependencies couldn't be built
cannot build derivation '/nix/store/fd4ab18rw7flk6s69lcz5cvg3lfwzbnx-markdown-0.1.17.4.drv': 13 dependencies couldn't be built
cannot build derivation '/nix/store/073ki4a4yjk32hhn3pj0kjwh6lmwgkhs-futhark-0.6.2.drv': 33 dependencies couldn't be built
error: build of '/nix/store/073ki4a4yjk32hhn3pj0kjwh6lmwgkhs-futhark-0.6.2.drv' failed
```
</p></details>
